README.md
Prerequisites:
1) Geth
2) Ganache
3) Truffle
4) Node.js
5) Metamask
6) IPFS API 

1) After you clone the repository, navigate to the repository folder and run "npm install"
2) Start ganache 
3) You will need 8 addresses for this project, mainly for 8 entitities listed as below:
    1] Journalist 1, Journalist 2
    2] Editor 1, Editor 2
    3] News Agency 1, News Agency 2
    4] Crowd Auditor 1, Crowd Auditor 2
4) Replace the public key in App.js (line 230 ) with the public key of Journalist 1 and Journalist 2 (according to your ganache public keys).
5) Replace the public key in App.js (line 248 ) with the public key of Editor 1 and Editor 2 (according to your ganache public keys).
6) Replace the public key in App.js (line 268 ) with the public key of News Agency 1 and News Agency 2 (according to your ganache public keys).
7) Replace the public key in Navbar.js (line 11) with the public key of Journalist 1 and Journalist 2 (according to your ganache public keys).
8) Replace the public key in Navbar.js (line 14) with the public key of Editor 1 and Editor 2 (according to your ganache public keys).
9) Replace the public key in Navbar.js (line 17) with the public key of News Agency 1 and News Agency 2 (according to your ganache public keys).
10) Replace the path for the images according to the respective paths in your device: Main.js[line 44] , Editor.js[line 75], NewsAgency.js [line 125], CrowdAuditor.js[line 111]
11) In truffle.config.js -> go to development{network{ change the port and host respective to your ganache}}
12) remove the build directory. 
13) navigate to the project folder in your terminal, run these commands:
      truffle compile
      truffle migrate --reset
      npm run start
14) If you still get errors then create a new project repository and then navigate to it in the terminal and type
      truffle init 
     Now go into the repository and copy the src folder, test folder, package.json file and repeat the steps from (3)


src/changelog.txt
Version: 2.1.1
  - Fixed animation on scroll effect not applying after portfolio filters clicked

Version: 2.1.0
  - Added Blog page
  - Dev version: prevent updating to jQuery version 3.5.0 since it's incompatible with the current Bootstrap version 

Version: 2.0.0
  - The template was rebuilt from scratch with the latest Bootstrap version (4.4.1)
  - Added SMPTP support for the contact form script (Pro)
  - Added NodeJS NPM Development version (Pro unlimited & Membership members)
  
Version: 1.0.0
  - Initial Release


test/SocialNetwork.js
const SocialNetwork = artifacts.require('./SocialNetwork.sol')

require('chai')
  .use(require('chai-as-promised'))
  .should()

contract('SocialNetwork', ([deployer, author, tipper, approver,crowdauditor1, crowdauditor2]) => {
  let socialNetwork
  
  before(async () => {
    socialNetwork = await SocialNetwork.deployed()
  })
   describe('deployment', async () => {
    it('deploys successfully', async () => {
     // socialNetwork = await SocialNetwork.deployed() 
      const address = await socialNetwork.address
      assert.notEqual(address, 0x0)
      assert.notEqual(address, '')
      assert.notEqual(address, null)
      assert.notEqual(address, undefined)
    })

    it('has a name', async() =>{
      const name = await socialNetwork.name()
      assert.equal(name, 'News Tracking and Tracing System')
    })
  })

  describe('posts', async() => {
    let result, postCount

    before(async () => {
      result = await socialNetwork.createPost('Test Article','This is my first post','abc123',{ from: author })
      postCount = await socialNetwork.postCount()
    })

    

    it('creates posts', async()=> {
    
     //Success
     assert.equal(postCount, 1)
     const event = result.logs[0].args
     assert.equal(event.id.toNumber(), postCount.toNumber(), 'id is correct')
     assert.equal(event.articleName, 'Test Article', 'Article name is correct' )
     assert.equal(event.content, 'This is my first post', 'content is correct')
     assert.equal(event.tipAmount, '0', 'tip amount is correct')
     assert.equal(event.approvalStatus, false, 'approvalStatus is correct')
     assert.equal(event.authenticity, true, 'The article is authentic')
     assert.equal(event.fakeCount,'0', 'The fakeCount is correct')
     assert.equal(event.trueCount,'0', 'The trueCount is correct')
     assert.equal(event.prevId, '1', 'The previous post id is correct')
     assert.equal(event.filehash, 'abc123', 'The filehash is correct')
     //Failure

     await socialNetwork.createPost('','','','', {from: author}).should.be.rejected;
    })

    it('lists posts', async()=> {
        const post = await socialNetwork.posts(postCount)
        assert.equal(post.id.toNumber(), postCount.toNumber(), 'id is correct')
        assert.equal(post.articleName, 'Test Article', 'Article name is correct' )
     assert.equal(post.content, 'This is my first post', 'content is correct')
     assert.equal(post.tipAmount, '0', 'tip amount is correct')
     assert.equal(post.approvalStatus, false, 'approvalStatus is correct')
     assert.equal(post.authenticity, true, 'The post is authentic')
     assert.equal(post.author, author, 'author is correct')
     assert.equal(post.filehash, 'abc123', 'The fielhash is correct')
    })

    it('allows users to tip posts', async()=> {
      const post = await socialNetwork.posts(postCount)

      let oldAuthorBalance
      oldAuthorBalance = await web3.eth.getBalance(author)
      oldAuthorBalance = new web3.utils.BN(oldAuthorBalance)

      result = await socialNetwork.tipPost(postCount, { from: tipper, value: web3.utils.toWei('1', 'Ether') })

      //Success
     const event = result.logs[0].args
     assert.equal(event.id.toNumber(), postCount.toNumber(), 'id is correct')
     assert.equal(event.articleName, 'Test Article', 'Article name is correct' )
     assert.equal(event.content, 'This is my first post', 'content is correct')
     assert.equal(event.tipAmount, '1000000000000000000', 'tip amount is correct')
    // assert.equal(event.approvalStatus, , 'approvalStatus is correct')
     assert.equal(event.author, author, 'author is correct')
     assert.equal(event.filehash, 'abc123', 'The fielhash is correct')

     //Check if the author received funds or not 
     let newAuthorBalance
     newAuthorBalance = await web3.eth.getBalance(author)
     newAuthorBalance = new web3.utils.BN(newAuthorBalance)

     let tipAmount
     tipAmount = web3.utils.toWei('1', 'Ether')
      tipAmount = new web3.utils.BN(tipAmount)

      const expectedBalance = oldAuthorBalance.add(tipAmount)
      assert.equal(newAuthorBalance.toString(), expectedBalance.toString())

      //failure: tries to tip a post that does not exist

      await socialNetwork.tipPost(99,{ from: tipper, value: web3.utils.toWei('1', 'Ether')}).should.be.rejected;
    })

     it('allows users to approve posts', async()=> {
       const post = await socialNetwork.posts(postCount)

      let oldApprovalStatus
      oldApprovalStatus = post.approvalStatus;

      result = await socialNetwork.approvePost(postCount,{ from: approver})

      //Success
     const event = result.logs[0].args
     assert.equal(event.id.toNumber(), postCount.toNumber(), 'id is correct')
     assert.equal(event.articleName, 'Test Article', 'Article name is correct' )
     assert.equal(event.content, 'This is my first post', 'content is correct')
     assert.equal(event.approvalStatus, true, 'approvalStatus is correct')
     assert.equal(event.authenticity, true, 'The post is authentic')
     assert.equal(event.author, author, 'author is correct')
     assert.equal(event.filehash, 'abc123', 'The fielhash is correct')

     //Check if the author received funds or not 
     let newApprovalStatus
     newApprovalStatus = post.approvalStatus

    let expectedApprovalStatus
    expectedApprovalStatus = post.approvalStatus

    assert.equal(expectedApprovalStatus,newApprovalStatus)
    
      //failure: tries to tip a post that does not exist

      await socialNetwork.tipPost(99,{ from: tipper, value: web3.utils.toWei('1', 'Ether')}).should.be.rejected;
    })

  it('allows users to report posts as fake', async()=> {
    const post = await socialNetwork.posts(postCount)

   let oldAuthenticity, oldCount
   oldAuthenticity = post.authenticity
   oldCount = post.fakeCount

   result = await socialNetwork.reportFake(postCount,{ from: crowdauditor1 || crowdauditor2})

   //Success
  const event = result.logs[0].args
  assert.equal(event.id.toNumber(), postCount.toNumber(), 'id is correct')
  assert.equal(event.articleName, 'Test Article', 'Article name is correct' )
  assert.equal(event.content, 'This is my first post', 'content is correct')
 // assert.equal(event.approvalStatus, true, 'approvalStatus is correct')
  assert.equal(event.authenticity, false, 'The post is reported fake')
  assert.equal(event.author, author, 'author is correct')
  assert.equal(event.filehash, 'abc123', 'The fielhash is correct')

  //Check if the author received funds or not 
  let newAuthenticity, newCount
  newAuthenticity = event.authenticity
  newCount = event.fakeCount

 let expectedAuthenticity, expectedCount
 expectedAuthenticity = !(oldAuthenticity)
 expectedCount = 1

 assert.equal(expectedAuthenticity.toString(),newAuthenticity.toString())
 assert.equal(expectedCount, newCount)
 
   //failure: tries to tip a post that does not exist

   await socialNetwork.reportFake(99,{ from: crowdauditor1 || crowdauditor2}).should.be.rejected;
 })

 it('allows users to report posts as true', async()=> {
  const post = await socialNetwork.posts(postCount)

 let oldAuthenticity, oldCount
 oldAuthenticity = post.authenticity
 oldCount = post.trueCount

 result = await socialNetwork.reportTrue(postCount,{ from: crowdauditor1 || crowdauditor2})

 //Success
const event = result.logs[0].args
assert.equal(event.id.toNumber(), postCount.toNumber(), 'id is correct')
assert.equal(event.articleName, 'Test Article', 'Article name is correct' )
assert.equal(event.content, 'This is my first post', 'content is correct')
// assert.equal(event.approvalStatus, true, 'approvalStatus is correct')
assert.equal(event.authenticity, true, 'The post is reported fake')
assert.equal(event.author, author, 'author is correct')
assert.equal(event.filehash, 'abc123', 'The fielhash is correct')

//Check if the author received funds or not 
let newAuthenticity, newCount
newAuthenticity = event.authenticity
newCount = event.trueCount.toString()

let expectedAuthenticity, expectedCount
expectedAuthenticity = !(oldAuthenticity)
expectedCount = oldCount++
expectedCount = 1

assert.equal(expectedAuthenticity.toString(),newAuthenticity.toString())
assert.equal(expectedCount, newCount)

 //failure: tries to tip a post that does not exist

 await socialNetwork.reportFake(99,{ from: crowdauditor1 || crowauditor2}).should.be.rejected;
})

it('modified posts', async()=> {
  //fetch the current post 
  const post = await socialNetwork.posts(postCount)
  // fetch the current post id and previous id
  let _id, oldprevId
  _id = post.id
  oldprevId = post.prevId
  //define new content
  const title = "modified article"
  const content = "modified content"
  const filehash = "abc123456"
  // modifying the post 
  result = await socialNetwork.modifyPost(_id,title,content,filehash,{from: tipper})
  postCount = await socialNetwork.postCount()
  //Success
  assert.equal(postCount,2)
  const event = result.logs[0].args
  assert.equal(event.id.toNumber(), postCount.toNumber(), 'id is correct')
  assert.equal(event.articleName, 'modified article', 'Article name is correct' )
  assert.equal(event.content, 'modified content', 'content is correct')
  assert.equal(event.tipAmount, '0', 'tip amount is correct')
  assert.equal(event.approvalStatus, false, 'approvalStatus is correct')
  assert.equal(event.authenticity, true, 'The article is authentic')
  assert.equal(event.fakeCount,'0', 'The fakeCount is correct')
  assert.equal(event.trueCount,'0', 'The trueCount is correct')
  assert.equal(event.prevId, '1', 'The previous post id is correct')
  assert.equal(event.filehash, 'abc123456', 'The fielhash is correct')

  //Failure
  
  await socialNetwork.modifyPost('','','', {from: author}).should.be.rejected;
 })

 it('retrieves the fileHash', async () => {
  const post = await socialNetwork.posts(postCount)
  let filehash = post.filehash
  const result = await socialNetwork.getHash(postCount)
  assert.equal(result, filehash,'get hash function works properly ')
  
})

})

describe('fakeArticles', async() => {
  let result1, fakeArticleCount, result, postCount

  before(async () => {
      result = await socialNetwork.createPost('Test Article','This is my first post','abc123',{ from: author })
      postCount = await socialNetwork.postCount()
    //fetch the current post 
  const post = await socialNetwork.posts(postCount)
  // fetch the current post id and previous id
   let _id
  _id = post.id
    result1 = await socialNetwork.addFakeArticle(_id)
    fakeArticleCount = await socialNetwork.fakeArticleCount()
  })

  it('add the post id to the fake articles list', async()=> {
    
    //Success
    assert.equal(postCount,3)
    assert.equal(fakeArticleCount,1)
    const event = result1.logs[0].args
    assert.equal(event.id.toNumber(), fakeArticleCount.toNumber(), 'id is correct')
    assert.equal(event.postid.toNumber(),postCount.toNumber(), 'Post id is correct' )
    
    //Failure
  
    await socialNetwork.addFakeArticle('').should.be.rejected;
   })
})

describe('trueArticles', async() => {
  let result1, trueArticleCount, result, postCount

  before(async () => {
      result = await socialNetwork.createPost('Test Article','This is my first post','abc123',{ from: author })
      postCount = await socialNetwork.postCount()
    //fetch the current post 
  const post = await socialNetwork.posts(postCount)
  // fetch the current post id and previous id
   let _id
  _id = post.id
    result1 = await socialNetwork.addTrueArticle(_id)
    trueArticleCount = await socialNetwork.trueArticleCount()
  })

  it('add the post id to the true articles list', async()=> {
    
    //Success
    assert.equal(postCount,4)
    assert.equal(trueArticleCount,1)
    const event = result1.logs[0].args
    assert.equal(event.id.toNumber(), trueArticleCount.toNumber(), 'id is correct')
    assert.equal(event.postid.toNumber(),postCount.toNumber(), 'Post id is correct' )
    
    //Failure
  
    await socialNetwork.addTrueArticle('').should.be.rejected;
   })
})


describe('addressStatus', async() => {
  let result1, addressCount, result, postCount

  before(async () => {
      result = await socialNetwork.createPost('Test Article','This is my first post','abc123',{ from: author })
      postCount = await socialNetwork.postCount()
    //fetch the current post 
  const post = await socialNetwork.posts(postCount)
  // fetch the current post id and previous id
   let _id
  _id = post.id
    result1 = await socialNetwork.addBlockedAddress(_id)
    addressCount = await socialNetwork.Addresscount()
  })

  it('add the address to a blocked address list', async()=> {
    
    //Success
    assert.equal(postCount,5)
    assert.equal(addressCount,1)
    const event = result1.logs[0].args
    assert.equal(event.id.toNumber(), addressCount.toNumber(), 'id is correct')
    assert.equal(event.postid.toNumber(),postCount.toNumber(), 'Post id is correct' )
    assert.equal(event.blocked,true, 'status is correct')
    
    //Failure
  
    await socialNetwork.addBlockedAddress('').should.be.rejected;
   })
})

})





src/components/Main.js
import React, { Component } from 'react';
import Identicon from 'identicon.js';
import './App.css';
import Breadcrumb from 'react-bootstrap/Breadcrumb';


const ipfsClient = require('ipfs-http-client')
const ipfs = ipfsClient({ host: 'ipfs.infura.io', port: '5001', protocol: 'https' })
var statement = "Upload Your File"

class Main extends Component {

  captureFile = (event) => {
    event.preventDefault();
    console.log("File Captured")
    const file = event.target.files[0]
    const reader = new window.FileReader() // converts the file to a buffer
    reader.readAsArrayBuffer(file)
    reader.onloadend = () => {
      this.setState({ buffer: Buffer(reader.result) })
      console.log('buffer', Buffer(reader.result))
    }

    const name = event.target.files[0].name;
    const lastDot = name.lastIndexOf('.');
    const uploadLabel = document.getElementById("uploadLabel");
    var fileName = name.substring(0, lastDot);
    const ext = name.substring(lastDot + 1);
    //fileName.concat(".");
    fileName.concat(ext);
    statement = fileName;
    uploadLabel.value = fileName;
  }

  render() {
    return (

      <div className="container-fluid">
<div class="bd-example " >
  <div id="carousel" class="carousel slide" data-ride="carousel">
   
    <div class="carousel-inner">
      <div class="carousel-item active">
        <img id="caroselimg" src={require("/home/akshada/social-network/src/components/images/image6.jpg")} class="d-block w-100" alt="Journalist Page"/>
        <div class="carousel-caption d-none d-md-block d-flex h-100 align-items-center justify-content-center">
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <br/>
          <font color="black">
          <b><h2>News Tracking and Tracing System Using Blockchain</h2></b>
          
          <h5>The News Tracking and Tracing System is built using</h5><h5> Ethereum, Ganache, Truffle and IPFS. 
          </h5>
          </font>
        </div>
        
      
      
      </div>
    </div>
    
  </div>
</div>
<br/>
<div className="row">

<main role="main" className="col-lg-12 ml-auto mr-auto">
<div class="card text-white bg-dark mb-3 col-lg-12 ml-auto mr-auto" style={{maxWidth: '700px'}}>
  <div class="card-header ml-auto mr-auto">Upload the News Article Here</div>
  <div class="card-body text-white ">
  <form onSubmit={(event) => {
                event.preventDefault()
                const content = this.postContent.value
                const Name = this.articleName.value
                console.log("Submitting the file..")
                ipfs.add(this.state.buffer, (error, result) => {
                  console.log("ipfs")
                  console.log('Ipfs result', result)
                  var filehash = result[0].hash
                  console.log(filehash)
                  
                  
                  this.setState({ filehash })
                  //this.setState({link})
                 // console.log(link)
                  if (error) {
                    console.log(error)
                    return
                  }
                  this.props.createPost(Name, content, filehash)
                })
              }
              }>
              
              <div className="form-group mr-sm-2 bg-transparent text-white" id="bg">
              <ul className="list-group list-group-flush text-white">  
                <li class="list-group-item text-white">
                  <input
                    id="articleName"
                    type="text"
                    ref={(input) => { this.articleName = input }}
                    className="form-control"
                    placeholder="Enter the article Name"
                    required />
                    </li>
                    <br/>
                  <li className="list-group-item text-white">
                  <input
                    id="postContent"
                    type="text"
                    ref={(input) => { this.postContent = input }}
                    className="form-control"
                    placeholder="Enter the article content"
                    required />
                    </li>
                    
                   <br/>
<li className="list-group-item text-white">
                  <div className="custom-file">
                    <input
                    id="fileinput"
                      type="file"
                      className="custom-file-input bg-dark"
                      onChange={this.captureFile}
                    />
                    <label className="custom-file-label bg-dark text-white" id="uploadLabel">
                      {statement}
                      </label>
                  </div>
                  </li>
                  </ul>
                </div>
                <button type="submit" className="btn btn-dark btn-outline-light btn-block">Upload</button>
              </form>
  
  </div>
</div>

       {/* <div className="row">

          <main role="main" className="col-lg-12 ml-auto mr-auto" style={{ maxWidth: '500px' }}>
            <div className="content mr-auto ml-auto">
              <form onSubmit={(event) => {
                event.preventDefault()
                const content = this.postContent.value
                const Name = this.articleName.value
                console.log("Submitting the file..")
                ipfs.add(this.state.buffer, (error, result) => {
                  console.log("ipfs")
                  console.log('Ipfs result', result)
                  var filehash = result[0].hash
                  console.log(filehash)
                  
                  
                  this.setState({ filehash })
                  //this.setState({link})
                 // console.log(link)
                  if (error) {
                    console.log(error)
                    return
                  }
                  this.props.createPost(Name, content, filehash)
                })
              }
              }>
                <div className="form-group mr-sm-2">
                  <input
                    id="articleName"
                    type="text"
                    ref={(input) => { this.articleName = input }}
                    className="form-control"
                    placeholder="Enter the article Name"
                    required />
                  <input
                    id="postContent"
                    type="text"
                    ref={(input) => { this.postContent = input }}
                    className="form-control"
                    placeholder="Enter the article content"
                    required />

                  <div className="custom-file">
                    <input
                      type="file"
                      className="custom-file-input"
                      onChange={this.captureFile}
                    />
                    <label className="custom-file-label" id="uploadLabel">
                      {statement}
                      </label>
                  </div>
                </div>
                <button type="submit" className="btn btn-primary btn-block">Upload</button>
              </form>*/
            }
             
  <div className="content mr-auto ml-auto">

              <br />
              {this.props.posts.map((post, key) => {
                var link="";
                link = "https://ipfs.infura.io/ipfs/"
                link =  link.concat(post.filehash)
                console.log("loop:", link)
                
                return (
                  <div class="card text-white bg-dark mb-3 col-lg-12 ml-auto mr-auto" key={key} id="cardDIV" style={{maxWidth: '700px'}}>
                  <div class="card-header ml-auto mr-auto">News Article {key}</div>
                  <div class="card-body text-white ">
                 {/* <div className="card mb-4" key={key} id="cardDIV" >
                    <div className="card-header">*/}
                    
                      Author  :
                      <img
                        className='mr-2'
                        width='30'
                        height='30'
                        alt="Identicon"
                        src={`data:image/png;base64,${new Identicon(post.author, 30).toString()}`}
                      />
                    
                     <small className="text-white">{post.author}</small>
                    <ul id="postList" className="list-group list-group-flush text-white">
                      <li className="list-group-item text-white">
                        <p>{post.articleName}</p>
                      </li>
                      <li className="list-group-item text-white">
                        <p>{post.content}</p>
                        
                        </li>
                      <li key={key} className="list-group-item py-2">
                        <small className="float-left mt-1 text-white">
                          TIPS: {window.web3.utils.fromWei(post.tipAmount.toString(), 'Ether')} ETH
                        </small>
                        <a className="btn btn-dark btn-outline-light float-right" href={link} role="button">File Link</a>
                      </li>
                    </ul>
                  </div>
                  </div>
        
                );
              })}
             
            </div>





          </main>
        </div>
      </div >
    );
  }
}

export default Main;

src/components/Editor.js
import React, { Component } from 'react';
import Identicon from 'identicon.js';
import './Editor.css'

const ipfsClient = require('ipfs-http-client')
const ipfs = ipfsClient({ host: 'ipfs.infura.io', port: '5001', protocol: 'https' })
var statement = "Upload Your File"
var Name = ""
var content = ""
var warning = "The article file has been modified"

class Editor extends Component {




  captureFile = (event) => {
    event.preventDefault();
    const file = event.target.files[0]
    const reader = new window.FileReader() // converts the file to a buffer
    reader.readAsArrayBuffer(file)
    reader.onloadend = () => {
      this.setState({ buffer: Buffer(reader.result) })
      console.log('buffer', Buffer(reader.result))
      //console.log("state buffer",this.state.buffer)
    }

    console.log("File Captured")

    var labelid = "uploadFile"
    var id = event.target.name
    id = id - 1
    console.log("IDDDDD:", id)
    labelid = labelid.concat(id)
    console.log("labelllll", labelid)
    const name = event.target.files[0].name;
    const lastDot = name.lastIndexOf('.');
    var uploadLabel = document.getElementById(labelid);
    var fileName = name.substring(0, lastDot);
    const ext = name.substring(lastDot + 1);
    fileName.concat(".");
    fileName.concat(ext);
    statement = fileName;
    uploadLabel.value = fileName;

  }

  warning(id) {
    id = id - 1
    console.log(id)
    console.log(this.props.posts[id])
    var post = this.props.posts[id]
    if (post.id !== post.prevId) {
      return (
        <li className="list-group-item-warning">
          The article file has been modified.
        </li>
      )
    }

  }

  render() {

    const imgurl = "/home/akshada/social-network/src/components/images/image6.jpg"
    // var filehash = "";

    return (
      <div className="container-fluid">
        <div class="bd-example " >
          <div id="carousel" class="carousel slide" data-ride="carousel">

            <div class="carousel-inner">
              <div class="carousel-item active">
                <img id="caroselimg" src={require("/home/akshada/social-network/src/components/images/image6.jpg")} class="d-block w-100" alt="Editor Page" />
                <div class="carousel-caption d-none d-md-block d-flex h-100 align-items-center justify-content-center">
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <br />
                  <font color="black">
                    <b><h2>News Tracking and Tracing System Using Blockchain</h2></b>

                    <h5>The News Tracking and Tracing System is built using</h5><h5> Ethereum, Ganache, Truffle and IPFS.
          </h5>
                  </font>
                </div>



              </div>
            </div>

          </div>
        </div>
        <br />
        <div className="row">
          <main
            role="main"
            className="col-lg-12 ml-auto mr-auto"

          >
            {this.props.posts.map((post, key) => {
              var link = "";
              link = "https://ipfs.infura.io/ipfs/"
              link = link.concat(post.filehash)
              var titleId = "title"
              titleId = titleId.concat(key)
              var contentId = "content"
              contentId = contentId.concat(key)
              var fileId = "file"
              fileId = fileId.concat(key)
              var labelId = "uploadFile"
              labelId = labelId.concat(key)

              return (
                <div class="card text-white bg-dark mb-3 col-lg-12 ml-auto mr-auto" key={key} id="cardDIV" style={{ maxWidth: '700px' }}>
                  <div class="card-header ml-auto mr-auto">News Article {key}</div>
                  <div class="card-body text-white ">
                    <form onSubmit={(event) => {
                      event.preventDefault()
                    }}>
                      Author:
                        <img
                        className='mr-2'
                        width='30'
                        height='30'
                        alt="Identicon"
                        src={`data:image/png;base64,${new Identicon(post.author, 30).toString()}`}
                      />
                      <small className="text-white">{post.author}</small>

                      <ul id="postList" className="list-group list-group-flush text-white">
                        {this.warning(key + 1)}
                        <li className="list-group-item text-white">
                          <input
                            id={titleId}
                            class="title text-white"
                            type="text"
                            ref={(input) => { this.articleName = input }}
                            value={this.props.value}
                            defaultValue={post.articleName}

                            required />


                        </li>
                        <li className="list-group-item text-white" >
                          <input
                            id={contentId}
                            type="text"
                            class="content"
                            ref={(input) => { this.postContent = input }}


                            defaultValue={post.content}

                            required />

                        </li>
                        
                          <a className="btn btn-dark btn-outline-light btn-block" href={link} role="button">Click here to see the File uploaded by the journalist</a>
                        
                        <li className="list-group-item py-2">
                          <div className="custom-file">
                            <input
                              type="file"
                              id={fileId}
                              name={post.id}
                              className="custom-file-input bg-dark"
                              onChange={this.captureFile}
                            />
                            <label className="custom-file-label bg-dark text-white" id={labelId} >
                              {statement}
                            </label>
                          </div>

                        </li>
                        <li key={key} className="list-group-item py-2">
                          <small className="float-left mt-1 text-white">
                            TIPS: {window.web3.utils.fromWei(post.tipAmount.toString(), 'Ether')} ETH
                        </small>

                          <button
                            id={key}
                            type="button"
                            className="btn btn-dark btn-outline-light float-right"
                            name={post.id}
                            onClick={(event) => {
                              var Name = ""
                              var content = ""
                              var postId = event.target.name
                              var keyy = postId - 1
                              console.log(keyy)
                              var title = "title"
                              var content = "content"
                              var file = "file"
                              title = title.concat(keyy)
                              content = content.concat(keyy)
                              file = file.concat(keyy)
                              console.log(title, content, file)
                              Name = document.getElementById(title).value
                              content = document.getElementById(content).value
                              console.log(Name);
                              console.log(content);
                              var id = event.target.name;
                              var filehash = "";
                              console.log(fileId)
                              console.log(file)
                              var file1 = document.getElementById(fileId).value
                              console.log(file1)
                              if (file1 != 0x0) {
                                console.log("file not empty")
                                console.log("Submitting the file..")
                                ipfs.add(this.state.buffer, (error, result) => {
                                  console.log('Ipfs result', result)
                                  filehash = result[0].hash
                                  console.log(filehash)
                                  this.setState({ filehash })
                                  console.log(this.state.filehash)
                                  this.setState({ link })
                                  this.props.modifyPost(id, Name, content, filehash)
                                  if (error) {
                                    console.log(error)
                                    return
                                  }
                                })
                              }
                              else {
                                console.log("file empty")
                                this.props.modifyPost(id, Name, content, post.filehash)
                              }


                            }}
                          >
                            Upload
                            </button>
                        </li>
                      </ul>

                    </form>
                  </div>
                </div>

              );
            })}

          </main>
        </div >
      </div>
    );
  }
}

export default Editor;